File: moxy-compiler/src/main/java/com/omegar/mvp/compiler/pipeline/TriplePublisher.java
package com.omegar.mvp.compiler.pipeline;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.Set;

/**
 * Created by Anton Knyazev on 05.12.2020.
 */
public class TriplePublisher<FI, SE, TH> extends Publisher<Triple<FI, SE, TH>> {
    private final List<FI> mFirsts = new ArrayList<>();
    private final List<SE> mSeconds = new ArrayList<>();
    private final List<TH> mThirds = new ArrayList<>();
    private final Publisher<FI> mPublisher1;
    private final Publisher<SE> mPublisher2;
    private final Publisher<TH> mPublisher3;


    public TriplePublisher(Publisher<FI> publisher1,
                           Publisher<SE> publisher2,
                           Publisher<TH> publisher3) {
        mPublisher1 = publisher1;
        mPublisher2 = publisher2;
        mPublisher3 = publisher3;
    }

    @Override
    public void publish(PipelineContext<Triple<FI, SE, TH>> context) {
        super.publish(context);

        mPublisher1.publish(new LocalContext<>(mFirsts));
        mPublisher2.publish(new LocalContext<>(mSeconds));
        mPublisher3.publish(new LocalContext<>(mThirds));
    }

    private void maybePublisherNext() {

        if (!mFirsts.isEmpty() && !mSeconds.isEmpty() && !mThirds.isEmpty() ) {
            Iterator<FI> firstIterator = mFirsts.iterator();
            Iterator<SE> secondIterator = mSeconds.iterator();
            Iterator<TH> thirdIterator = mThirds.iterator();

            while (firstIterator.hasNext() && secondIterator.hasNext() && thirdIterator.hasNext() ) {
                FI first = firstIterator.next();
                SE second = secondIterator.next();
                TH third = thirdIterator.next();

                next(new Triple<>(first, second, third));
                firstIterator.remove();
                secondIterator.remove();
                thirdIterator.remove();
            }
        }
    }

    public static <FI, SE, TH> TriplePublisher<Set<FI>, Set<SE>, Set<TH>> collectQuad(Publisher<FI> publisher1,
                                                                                                   Publisher<SE> publisher2,
                                                                                                   Publisher<TH> publisher3) {
        return new TriplePublisher<>(
                publisher1.collect(),
                publisher2.collect(),
                publisher3.collect()
        );
    }

    private class LocalContext<T> implements PipelineContext<T> {

        private final List<T> list;

        public LocalContext(List<T> list) {
            this.list = list;
        }

        @Override
        public void next(T nextData) {
            list.add(nextData);
            maybePublisherNext();
        }

        @Override
        public void finish() {
            // nothing
        }
    }


}

File: moxy/compiler/src/main/java/com/omegar/mvp/compiler/pipeline/TriplePublisher.kt
package com.omegar.mvp.compiler.pipeline

/**
 * Created by Anton Knyazev on 05.12.2020.
 */
class TriplePublisher<FI, SE, TH>(
    private val mPublisher1: Publisher<FI>,
    private val mPublisher2: Publisher<SE>,
    private val mPublisher3: Publisher<TH>
) : Publisher<Triple<FI, SE, TH>>() {

    private val mFirsts: MutableList<FI> = ArrayList()
    private val mSeconds: MutableList<SE> = ArrayList()
    private val mThirds: MutableList<TH> = ArrayList()

    override fun publish(context: PipelineContext<Triple<FI, SE, TH>>?) {
        super.publish(context)
        mPublisher1.publish(LocalContext(mFirsts))
        mPublisher2.publish(LocalContext(mSeconds))
        mPublisher3.publish(LocalContext(mThirds))
    }

    private fun maybePublisherNext() {
        if (mFirsts.isNotEmpty() && mSeconds.isNotEmpty() && mThirds.isNotEmpty()) {
            val firstIterator = mFirsts.iterator()
            val secondIterator = mSeconds.iterator()
            val thirdIterator = mThirds.iterator()
            while (firstIterator.hasNext() && secondIterator.hasNext() && thirdIterator.hasNext()) {
                val first = firstIterator.next()
                val second = secondIterator.next()
                val third = thirdIterator.next()
                next(Triple(first, second, third))
                firstIterator.remove()
                secondIterator.remove()
                thirdIterator.remove()
            }
        }
    }

    private inner class LocalContext<T>(private val list: MutableList<T>) : PipelineContext<T> {
        override fun next(nextData: T) {
            list.add(nextData)
            maybePublisherNext()
        }

        override fun finish() {
            // nothing
        }
    }

    companion object {
        @JvmStatic
        fun <FI, SE, TH> collectTriple(
            publisher1: Publisher<FI>,
            publisher2: Publisher<SE>,
            publisher3: Publisher<TH>
        ): TriplePublisher<Set<FI>, Set<SE>, Set<TH>> {
            return TriplePublisher(
                publisher1.collect(),
                publisher2.collect(),
                publisher3.collect()
            )
        }
    }
}